### PR TITLE
4838 d contract expiry specs

### LINKED ISSUE #4840: If a fungible has custom fees, and the contract is deleted at end of grace period, will the fungible balance be transferred back to treasury without charging the account?
- [ ] #4847

### LINKED ISSUE #4841: If an NFT has custom fees, and the contract is deleted at end of grace period, will the NFT be transferred back to treasury without charging the account?
- [ ] #4847

### LINKED ISSUE #4845: Also verify if the treasury has "receiver signature required" is bypassed in the case of expiry when deleted at the end of the grace period

### LINKED ISSUE #4847: Add a test to validate storage expiration works at expected interval, when doing transaction with custom fees

### PR DESCRIPTION
**Pull request by kimbor**
Fixes #4840
Fixes #4841 
Fixes #4847
Fixes #4845

### CODE REVIEW

File: test-clients/src/main/java/com/hedera/services/bdd/suites/autorenew/ContractAutoExpirySpecs.java [hunk 1]
```diff
@@ -391,22 +403,29 @@ private HapiSpec storageExpiryWorksAtTheExpectedInterval() {
...
                         cryptoTransfer(
                                 moving(aFungibleAmount, aFungibleToken)
                                         .between(TOKEN_TREASURY, contractToRemove),
                                 moving(bFungibleAmount, bFungibleToken)
                                         .between(TOKEN_TREASURY, contractToRemove),
+                                moving(cFungibleAmount, cFungibleTokenWithCustomFees)
+                                        .between(TOKEN_TREASURY, contractToRemove),
                                 movingUnique(nonFungibleToken, 1L, 2L)
                                         .between(TOKEN_TREASURY, contractToRemove)),
+                        // TODO: would like to add a check that the token move was successful
```
Comment (kimbor) [inline test-clients/src/main/java/com/hedera/services/bdd/suites/autorenew/ContractAutoExpirySpecs.java:428]: Is it possible / desirable to add this check? It seems that in order to check this, we'd have to wait for another round of consensus, significantly slowing this test down.

Comment (Neeharika-Sompalli) [reply test-clients/src/main/java/com/hedera/services/bdd/suites/autorenew/ContractAutoExpirySpecs.java:429]: You mean submitting another transaction is slowing down the test ?

**Review by povolev15 — DISMISSED** [review]
LGTM

File: test-clients/src/main/java/com/hedera/services/bdd/suites/autorenew/ContractAutoExpirySpecs.java (hunk 1 above)
Comment (kimbor) [reply test-clients/src/main/java/com/hedera/services/bdd/suites/autorenew/ContractAutoExpirySpecs.java]: I added a check that the token move was successful, as we discussed.

File: test-clients/src/main/java/com/hedera/services/bdd/suites/autorenew/ContractAutoExpirySpecs.java [hunk 2]
```diff
@@ -606,6 +608,77 @@ private HapiSpec renewsUsingContractFundsIfNoAutoRenewAccount() {
...
+                                moving(aFungibleAmount, aFungibleToken)
+                                        .between(TOKEN_TREASURY, contractToRemove),
+                                movingUnique(nonFungibleToken, 1L, 2L)
+                                        .between(TOKEN_TREASURY, contractToRemove)),
+                        getAccountBalance(contractToRemove)
+                                .hasTokenBalance(aFungibleToken, aFungibleAmount)
+                                .hasTokenBalance(nonFungibleToken, 2),
+                        getAccountBalance(TOKEN_TREASURY).hasTokenBalance(aFungibleToken, 0),
+                        getAccountInfo(TOKEN_TREASURY).hasOwnedNfts(0),
+                        /* sleep past the expiration:
```
Comment (kimbor) [inline test-clients/src/main/java/com/hedera/services/bdd/suites/autorenew/ContractAutoExpirySpecs.java:896]: This comment is helpful, thanks

File: test-clients/src/main/java/com/hedera/services/bdd/suites/autorenew/ContractAutoExpirySpecs.java [hunk 3]
```diff
@@ -606,6 +608,77 @@ private HapiSpec renewsUsingContractFundsIfNoAutoRenewAccount() {
...
+                        // check that the contract is gone
+                        getContractInfo(contractToRemove)
+                                .hasCostAnswerPrecheck(INVALID_CONTRACT_ID),
+                        // And the fungible units were returned to the treasury
+                        getAccountBalance(TOKEN_TREASURY)
+                                .hasTokenBalance(aFungibleToken, aFungibleAmount),
+                        // And the NFTs are now owned by the treasury
+                        getTokenNftInfo(nonFungibleToken, 1L).hasAccountID(TOKEN_TREASURY),
+                        getTokenNftInfo(nonFungibleToken, 2L).hasAccountID(TOKEN_TREASURY)
+                        // TODO: re-enable after expiry throttling is fixed
```
Comment (kimbor) [inline test-clients/src/main/java/com/hedera/services/bdd/suites/autorenew/ContractAutoExpirySpecs.java:917]: TODO

File: test-clients/src/main/java/com/hedera/services/bdd/suites/autorenew/ContractAutoExpirySpecs.java [hunk 4]
```diff
@@ -658,10 +684,77 @@ private HapiSpec storageExpiryWorksAtTheExpectedInterval() {
...
+
+                        // And the NFTs are now owned by the treasury
+                        getTokenNftInfo(nonFungibleToken, 1L).hasAccountID(TOKEN_TREASURY),
+                        getTokenNftInfo(nonFungibleToken, 2L).hasAccountID(TOKEN_TREASURY),
+                        // And the account was not charged for the transfer
+                        getAccountBalance(FEE_COLLECTOR).hasTinyBars(ONE_HBAR),
+                        getAccountBalance(TOKEN_TREASURY).hasTinyBars(ONE_HBAR));
+    }
+
+    private HapiSpec verifyNonFungibleTokenTransferredBackToTreasuryWithoutCharging() {
```
[the remark this answers is not in the document]
Comment (povolev15) [reply test-clients/src/main/java/com/hedera/services/bdd/suites/autorenew/ContractAutoExpirySpecs.java:705]: I want that it will be explicitly for NFT

[the remark this answers is not in the document]
Comment (kimbor) [reply test-clients/src/main/java/com/hedera/services/bdd/suites/autorenew/ContractAutoExpirySpecs.java:705]: There are already some NFT tests in storageExpiryWorksAtTheExpectedInterval

[the remark this answers is not in the document]
Comment (povolev15) [reply test-clients/src/main/java/com/hedera/services/bdd/suites/autorenew/ContractAutoExpirySpecs.java:705]: you are right but that one is explicitly for NFT and not for other. why to combine all test in one test?

File: test-clients/src/main/java/com/hedera/services/bdd/suites/autorenew/ContractAutoExpirySpecs.java (hunk 3 above)
Comment (mhess-swl) [reply test-clients/src/main/java/com/hedera/services/bdd/suites/autorenew/ContractAutoExpirySpecs.java]: Issue fixed, `TODO` no longer needed. I removed the `TODO` comment

**Review by mhess-swl — DISMISSED** [review]

File: hedera-node/test-clients/src/main/java/com/hedera/services/bdd/suites/autorenew/ContractAutoExpirySpecs.java [hunk 5]
```diff
@@ -92,7 +92,7 @@ public static void main(String... args) {
 
     @Override
     public List<HapiSpec> getSpecsInSuite() {
-        return List.of(
+        return List.of(new HapiSpec[] {
```
Comment (kimbor) [inline hedera-node/test-clients/src/main/java/com/hedera/services/bdd/suites/autorenew/ContractAutoExpirySpecs.java:95]: Matt, why did you add this array creation back in? I had taken it out because SonarLint was complaining about it. List.of(_) can take the elements directly as arguments, instead of passing an array.

File: test-clients/src/main/java/com/hedera/services/bdd/suites/autorenew/ContractAutoExpirySpecs.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (kimbor) [reply test-clients/src/main/java/com/hedera/services/bdd/suites/autorenew/ContractAutoExpirySpecs.java]: Ok to leave this as a separate test, per our discussion in Slack

**Comment by kimbor** [comment]
I made a lot of changes to appease SonarCloud. The only error (non-warning) that it's complaining about now is the presence of assert statements in the EET test class ContractAutoExpirySpecs, which it doesn't like in non-test code. It decides whether a given class is test based on whether the word 'test' is present in the package name, which it isn't, so it thinks this class is production code. So basically this is a bogus error which we should ignore.

**Review by povolev15 — DISMISSED** [review]
LGTM

**Review by Neeharika-Sompalli — COMMENTED** [review]
I think `hedera-node/hapi/*` files are deleted , which should not be deleted .

File: hedera-node/test-clients/src/main/java/com/hedera/services/bdd/suites/autorenew/ContractAutoExpirySpecs.java (hunk 5 above)
Comment (mhess-swl) [reply hedera-node/test-clients/src/main/java/com/hedera/services/bdd/suites/autorenew/ContractAutoExpirySpecs.java]: Sorry. It's easier to comment out tests individually with an array because the last array element can be followed by a comma, which takes about a third of the three seconds it takes to move a parenthesis

**Comment by povolev15** [comment]
> 

I reverted these files

**Review by mhess-swl — APPROVED** [review]

**Review by povolev15 — APPROVED** [review]
LGTM

**Review by nathanklick — APPROVED** [review]
